### ISSUE TITLE
Allow users to define their own i18n (e.g. for button value text) 

### ISSUE THREAD

**Issue by zhengxiaoyao0716** [gradio-app/gradio#2465]
### Describe the bug

See [Button.svelte#L14](https://github.com/gradio-app/gradio/blob/5e7cf62a478b2eeea8026d2e2ae94de7811b6504/ui/packages/app/src/components/Button/Button.svelte#L14), the value of the `Button` components has always try to translated by `svelte-i18n`, but I have not found the way to custom the locale dictionaries, that caused all the button value except defined under the `/ui/packages/app/src/lang` would make a browser warning gave by `svelte-i18n`.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

https://github.com/AUTOMATIC1111/stable-diffusion-webui

### Screenshot

![image](https://user-images.githubusercontent.com/12682063/195916701-ce0894f6-a1f8-4c1d-893b-7257f5172ddb.png)

### Logs

```shell
browser warning, without python console log.
```

### System Info

```shell
gradio==3.4.1
```

### Severity

blocking all usage of gradio

**Comment by pngwn** [comment]
Is the issue here that you wish to get rid of the console logs or that you wish to set your own language definitions?

**Comment by zhengxiaoyao0716** [comment]
Set custom language definitions feels better, but it means maybe we should make all components used `sevlte-i18n` to show the custom values, not only the `Button` component.

**Comment by abidlabs** [comment]
Renamed and categorized as an enhancement

**Comment by abidlabs** [comment]
Hi @zhengxiaoyao0716 it is now possible to read a user's browser locale and change your function / UI based on that. The way you do that is by passing in a `gr.Request` object into your function which contains the request headers. It's probably easiest to explain by an example, so let me provide two:

**Example 1: changing a function's output based on user language**

```py
import gradio as gr

def greet(name, request: gr.Request):
    if request.headers["Accept-Language"].split(",")[0].lower().startswith("en"):
        return "Hello, " + name
    elif request.headers["Accept-Language"].split(",")[0].lower().startswith("es"):
        return "Hola, " + name
    else:
        return "Language not recognized"

gr.Interface(greet, "text", "text").launch()
```

Which produces (if you have set Spanish in your browser): 

<img width="578" alt="image" src="https://user-images.githubusercontent.com/1778297/221762792-cc747054-d5bf-4763-a1fc-19497b89fd52.png">

You can also change button values, like you asked:

**Example 2: changing a button value based on user language**

```py
import gradio as gr

def greet(request: gr.Request):
    if request.headers["Accept-Language"].split(",")[0].lower().startswith("en"):
        return "Hello"
    elif request.headers["Accept-Language"].split(",")[0].lower().startswith("es"):
        return "Hola"
    else:
        return "Language not recognized"


with gr.Blocks() as demo:
    btn = gr.Button()
    demo.load(greet, None, btn)
    
demo.launch()
```

Which produces (if you have set Spanish in your browser): 

<img width="583" alt="image" src="https://user-images.githubusercontent.com/1778297/221763046-4cb66b45-60d7-467c-a475-6a6e709699d7.png">

Thanks for raising this issue!

**Comment by freddyaboulton** [comment]
@abidlabs Very cool workaround. I think we should reopen this issue based on the recent discussions regarding localizing text between Chinese and English for https://huggingface.co/spaces/JohnSmith9982/ChuanhuChatGPT ?

**Comment by abidlabs** [comment]
Sounds good @freddyaboulton! Basically, we want to provide a simpler API (and one that ideally does not need to make requests every time to the server) for this, right?

**Comment by pngwn** [comment]
Fundamentally i think the solution should be something like, any 'value' can accept a string or an instance of `gr.I18n()`/ `gr.Translate()` which can provide a series of values to be chosen by the i18n tools we use on the frontend.

Something like:

```py
gr.translate(en = 'Hello', es = 'Hola')
```

I've lost the dictionary battle before, however in this instance I think it actually _should_ take a dictionary because most i18n standards are built around JSON and it isn't unreasonable for user to generate their translation string elsewhere (maybe even with a different tool/ gui tool) export them and import them into a gradio app. 

A user needs to be able to do this:

```py
import translation # some dict that used to be a json file or something

gr.Translate(translation)
```

`gr.Translation` could be passed in to Blocks and provide _additional_ translations or we could support translations directly on the component values themselves. We could also support both, with component specific values taking priority where there are collisions.

In order to support this we need to wrap all string with i18n helpers on the frontend. Because the top level version would actually allow a user to provide translations for something that we may not have full translations for but that are not a component value.

**Comment by abidlabs** [comment]
I agree that a dict structure makes sense here @pngwn! We don't want to hardcode a set of parameters for every language, so something like:

`gr.Translation({"en": "hello", "es": "hola"})` 

makes sense to me. 

Some additional thoughts:

* Should we support this for every component value (`Dataframe`), or only those that take strings (like `Button`, `Textbox`). I think only strings is good for now
* This should be able to be returned by a function as well, not just the initial value

Implementation-wise,

`gr.Translation({"en": "hello", "es": "hola"})`  could return a plain dictionary with a special key that the frontend internationalizes. This way, we could make it explicitly listed in the config so that no additional server requests are needed to fetch the internationalized values.

**Comment by freddyaboulton** [comment]
I was thinking this would only apply to labels, info text, button text, etc and not the arbitrary values of the components. 

My thought process was that the LLMs and other models that app authors interface with support multiple languages already. So it's not that the input/output of the model needs to be translated, but rather the text displayed in the app to guide a user on how to use the app.

Example from ChuanhuChatGPT supporting english but the app text is in chinese:

<img width="1367" alt="image" src="https://user-images.githubusercontent.com/41651716/227567030-558550fb-3850-4255-8fe5-9ac07d993ff3.png">

**Comment by Sakura-Luna** [comment]
How can I disable i18n for projects that don't use i18n for localization?

**Comment by freddyaboulton** [comment]
I'm not sure that's possible right now @Sakura-Luna. Tagging @pngwn @dawoodkhan82 and @aliabid94 to confirm.

**Comment by Sakura-Luna** [comment]
I hope at least this frequent error can be resolved.
![image](https://user-images.githubusercontent.com/53183413/232088851-d4047add-4346-4eb3-9a98-e860d9f7de6f.png)

**Comment by stunlocked1** [comment]
> Hi @zhengxiaoyao0716 it is now possible to read a user's browser locale and change your function / UI based on that. The way you do that is by passing in a `gr.Request` object into your function which contains the request headers. It's probably easiest to explain by an example, so let me provide two:
> 
> **Example 1: changing a function's output based on user language**
> 
> ```python
> import gradio as gr
> 
> def greet(name, request: gr.Request):
>     if request.headers["Accept-Language"].split(",")[0].lower().startswith("en"):
>         return "Hello, " + name
>     elif request.headers["Accept-Language"].split(",")[0].lower().startswith("es"):
>         return "Hola, " + name
>     else:
>         return "Language not recognized"
> 
> gr.Interface(greet, "text", "text").launch()
> ```
> 
> Which produces (if you have set Spanish in your browser):
> 
> <img alt="image" width="578" src="https://user-images.githubusercontent.com/1778297/221762792-cc747054-d5bf-4763-a1fc-19497b89fd52.png">
> 
> You can also change button values, like you asked:
> 
> **Example 2: changing a button value based on user language**
> 
> ```python
> import gradio as gr
> 
> def greet(request: gr.Request):
>     if request.headers["Accept-Language"].split(",")[0].lower().startswith("en"):
>         return "Hello"
>     elif request.headers["Accept-Language"].split(",")[0].lower().startswith("es"):
>         return "Hola"
>     else:
>         return "Language not recognized"
> 
> 
> with gr.Blocks() as demo:
>     btn = gr.Button()
>     demo.load(greet, None, btn)
>     
> demo.launch()
> ```
> 
> Which produces (if you have set Spanish in your browser):
> 
> <img alt="image" width="583" src="https://user-images.githubusercontent.com/1778297/221763046-4cb66b45-60d7-467c-a475-6a6e709699d7.png">
> 
> Thanks for raising this issue!

how did you set buttons to spanish? It says Limpiar Enviar, but for me buttons stay on english

**Comment by freddyaboulton** [comment]
I had to change the system language on my computer. But there may be a way to change the language locale in your browser.

**Comment by pngwn** [comment]
You. A set the preferred language in the browser settings iirc.

**Comment by branaway** [comment]
how to set the global language setting such that the image component uses the correct built-in localized string?
![Screen Shot 2024-05-30 at 18 17 44](https://github.com/gradio-app/gradio/assets/179513/032dbec9-c559-4cb1-8298-dd9d69d43bb5)

**Comment by freddyaboulton** [comment]
Wow awesome @hoveychen !

**Comment by branaway** [comment]
Nice feature to have!
